keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md

```diff
@@ -0,0 +1,163 @@
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposal
+
+The NodeRestriction admission plugin will be updated to limit the labels a kubelet
+can set on its Node object. Kubelets will only be allowed to set the following labels.
+
+### Well-known label keys
+
+This list includes all of the labels that kubelets already set on their Node objects.
+
+```
+kubernetes.io/hostname
+failure-domain.beta.kubernetes.io/zone
+failure-domain.beta.kubernetes.io/region
+beta.kubernetes.io/instance-type
+beta.kubernetes.io/os
+beta.kubernetes.io/arch
+```
+
+### Well-known kubelet-related label namespace
+
+This prefix allows kubelets to control arbitrary labels under their own label namespace.
+
+```
+[*.]kubelet.kubernetes.io/*
+```
+
+### Well-known topology-related label namespace
+
+This prefix allows kubelets to control arbitrary labels related to topology.
+
+```
+[*.]topology.kubernetes.io/*
+```
+
+#### CSI use case
+
+CSI allows drivers to provide [arbitrary topology labels](https://github.com/container-storage-interface/spec/blob/v0.3.0/csi.proto#L524-L549),
+which are planned to be added as labels to the Node by the kubelet.
+This section describes how CSI could continue to support arbitrary topology information provided by drivers,
+and conform to a consistent topology-related prefix when adding labels to Nodes.
+
+The CSI topology labels are in the form `<name>` or `<prefix>/<name>`, where both prefix and name are up to 63 characters long.
+
+Kubernetes labels allow a prefix up to [253 characters long](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+
+CSI topology labels could be transformed to kubernetes labels within the permitted topology label namespace:
+
+* CSI topology labels that did not have a prefix would have a prefix of `topology.kubernetes.io/` prepended
+* CSI topology labels with a prefix not ending in `topology.kubernetes.io` would have that string added to their prefix
+
+Component behavior:
+
+* The kubelet would transform CSI topology to kubernetes labels before adding them to the `Node` API object.
+* The CSI provisioner would transform CSI topology labels to kubernetes labels when [retrieving topology info from a given node](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305).
```

Comment:
The other way around? i.e. from Kubernetes labels to CSI topology labels in order to generate the right CSI accessibility requirements

Reply:
I think this is correct… the input is topology keys, and it needs to look up the values from the node labels, so it translates from CSI topology keys to label keys

Reply:
right, but then the output should be converted from k8s label format to CSI topology format so it could be processed as required/preferred topology

Reply:
The translated key is only used to look up the label value from the node object: https://github.com/kubernetes-csi/external-provisioner/pull/160/commits/0f111a2c50aa4a58a139f9d5314fecad3f7af9bf#diff-94bad98230836caa3e44ca0c9b4281c3L294